### DISCUSSION TITLE
Rename file with current date: Problem with cd'ing into current directory

### DISCUSSION THREAD

**Discussion by ReaderGuy42** [Taitava/obsidian-shellcommands#369] [Q&A] [answered]
I'd like to rename file `X` into `current-date X`. I can do this if I specify the concrete directory the file is in, e.g.
```
gnome-terminal -- bash -c "cd /Documents/Obsidian/Books/0\ Currently\ Reading && mv {{file_name}} /Documents/Obsidian/Booklog/{{date:YYYY}}/{{date:MM}}/{{date:YYYY-MM-DD}}\ {{file_name}} && exit; exec bash"
```
This moves the active file from `0 Currently Reading` into `Booklog`, into the current year folder and the current month folder, so e.g. `Booklog/2023/09/X` with X being the book note. This works.

However, if I want to do this to a note that isn't in the specified folder, this won't work because obviously the `cd` command goes to that particular folder and the note `Y` isn't in there. And this also doesn't work if you want to rename a file but that file should remain where it is.

```
gnome-terminal -- bash -c "mv {{file_path:absolute}} {{date:YYYY-MM-DD}}\ {{file_name}} && exit; exec bash"
```
This will rename the file correctly but will put it in the root directory of the vault. However, I can't figure out how to rename the file in place.  Because with `{{file_path:absolute}}` (or `relative`) you can't `cd` into that directory because it's actually describing the file path including the file.

So the main thing would be to have a `{{file_path}}` type variable that describes the directory but doesn't include the file itself. 

Is this possible?

My goal is to eventually have a flexible shell script that would check if the file includes a string, e.g. `Medium: Book` and then moves it to one folder, or if it has another string, e.g. `Medium: Movie` and moves it to another folder.

But this is based on being `a)` being able to rename something in the same place it is (this is also required for notes that are already in the correct place and I just want to rename) and `b)` being able to tell bash where to look for the file without having to manually give it the exact address.

Thanks!

EDIT: Just discovered `{{folder_path:absolute}}`. Playing around with that, I'll come back and amend this once I've figured that out.

EDIT 2:
So, this works:
```
gnome-terminal -- bash -c "case {{yaml_value:Medium}} in Book | Audiobook | Podcast | eBook | GraphicNovel) mv {{file_path:absolute}}  {{vault_path}}/Book\ Log/{{date:YYYY}}/{{date:MM}}/{{date:YYYY-MM-DD}}\ {{file_name}} ;; Movie | TV | Game) mv {{file_path:absolute}}  {{vault_path}}/Movies\ and\ Games/{{date:YYYY}}/{{date:MM}}/{{date:YYYY-MM-DD}}\ {{file_name}};; esac; exit; exec bash"
```
and does exactly what I wanted it to. It uses `case` because that's what I got to work first. It checks what the `Medium` is in the front matter of my notes and sorts them accordingly.

So now I'm still working on an if-clause that checks if the file being looked at doesn't already have a date in it, so it doesn't get two.

Just aesthetically/visually, it would be easier if your plugin could be written with line breaks in them. Those seem to make the scripts unusable so it seems for now I'm stuck with one-liners. But, as I have no idea how difficult this would be to change, that's fine.

What's more important: how do I tell the shell script to stay with the current note? When I have the above script move my file, Obsidian then opens another note (I can't tell why it chooses that other note). So `X` is active, I trigger the shell command to move the file, but `X` should remain active. How can I do this?

Thanks :)

**Comment by ReaderGuy42** [comment]
This is what I currently have for the check if there's a date in the file:
```
gnome-terminal -- bash -c 

"case 
{{file_name}} != 20*) 
	case {{yaml_value:Medium}} in 
		Book | Audiobook | Podcast | eBook | GraphicNovel) 
			mv {{file_path:absolute}}  {{vault_path}}/Book\ Log/{{date:YYYY}}/{{date:MM}}/{{date:YYYY-MM-DD}}\ {{file_name}} ;; 
		Movie | TV | Game) 
			mv {{file_path:absolute}} {{vault_path}}/Movies\ and\ Games/{{date:YYYY}}/{{date:MM}}/{{date:YYYY-MM-DD}}\ {{file_name}} ;; 
	esac ;; 
		
{{file_name}} == 20*)
	 case {{yaml_value:Medium}} in 
	 	Book | Audiobook | Podcast | eBook | GraphicNovel) 
	 		mv {{file_path:absolute}}  {{vault_path}}/Book\ Log/{{date:YYYY}}/{{date:MM}}/{{file_name}} ;; 
	 	Movie | TV | Game) 
	 		mv {{file_path:absolute}} {{vault_path}}/Movies\ and\ Games/{{date:YYYY}}/{{date:MM}}/{{file_name}} ;; 
	 esac ;; 

esac	 
exec bash"
```
This does NOT work. A terminal window pops up but closes right away, so I don't get any error message. This is just for clarity, for the plugin, this needs to be turned into a one-liner. If anyone has any ideas what's breaking this, I'd appreciate it :)

**Reply by Taitava** [reply]
I just [answered below](https://github.com/Taitava/obsidian-shellcommands/discussions/369#discussioncomment-7036550), but I'll mention this here, too: You can try to add a backslash `\` at the end of each line that is continued on the next line.

**Comment by ReaderGuy42** [comment]
I think this is working now:
```
case {{file_name}} in 
    2*) : ;; 
    *) mv {{file_path:absolute}} {{folder_path:absolute}}/{{date:YYYY-MM-DD}}\ {{file_name}} ;;  
esac; 

case {{yaml_value:Medium}} in 
Book | Audiobook | Podcast | eBook | GraphicNovel) mv {{file_path:absolute}} {{vault_path}}/Book\ Log/{{date:YYYY}}/{{date:MM}}/{{file_name}} ;; 
Movie | TV | Game) mv {{file_path:absolute}}  {{vault_path}}/Movies\ and\ Games/{{date:YYYY}}/{{date:MM}}/{{file_name}};; 

esac; exit; 

exec bash
```
I tried using negations e.g. ![2*] for the case that the file doesn't start with a number but this doesn't seem to work in the plugin, so "*" appears to match everything not matched by the previous case? Not sure I understood the tutorial correctly. If this doesn't work, I'll replace it with `a*|b*|c*` etc.

As a one-liner for the plugin this looks like this:
```
gnome-terminal -- bash -c "case {{file_name}} in 2*) : ;; *) mv {{file_path:absolute}} {{folder_path:absolute}}/{{date:YYYY-MM-DD}}\ {{file_name}} ;; esac; case {{yaml_value:Medium}} in Book | Audiobook | Podcast | eBook | GraphicNovel) mv {{file_path:absolute}}  {{vault_path}}/Book\ Log/{{date:YYYY}}/{{date:MM}}/{{file_name}} ;; Movie | TV | Game) mv {{file_path:absolute}}  {{vault_path}}/Movies\ and\ Games/{{date:YYYY}}/{{date:MM}}/{{file_name}};; esac; exit; exec bash"
```
If anyone has any improvements, please let me know :)

**Reply by ReaderGuy42** [reply]
OK; so this version works in renaming the file. If and only if the file already has a date at the start it moves it, otherwise it just renames it in place. I'm still working on a `first rename then move` thing.

**Reply by Taitava** [reply]
Sorry, I'm not familiar with `case`.

**Comment by ReaderGuy42** [comment]
I've now rewritten this in regular bash, and am now (so far unsuccessfully) trying to just have Shellcommands run it.
The following `.sh` works as desired if I run it directly in terminal. I can run `bash rename-move.sh test.md` and it renames and moves the file as I set out to do.

```
#!/bin/bash

fileName="$1"
f10="${fileName:0:10}"
today=$(date +"%Y-%m-%d")
year=$(date +"%Y")
month=$(date +"%m")


if  [[ $f10 != [0-9]* ]] && grep -qE '(Medium: Book|Medium: Audiobook|Medium: Podcast|Medium: eBook|Medium: GraphicNovel)' "$1"; then

        mv "$1" "/home/Documents/Obsidian/Books and Movies/Book Log/$year/$month/$today $1"

elif [[ $f10 != [0-9]* ]] && grep -qE '(Medium: Movie|Medium: TV|Medium: Game)' "$1"; then
        
        mv "$1" "/home/Documents/Obsidian/Books and Movies/Movies and Games/$year/$month/$today $1"

fi

if [[ $f10 = [0-9]* ]]  && grep -qE '(Medium: Book|Medium: Audiobook|Medium: Podcast|Medium: eBook|Medium: GraphicNovel)' "$1"; then

        mv "$1" "/home/Documents/Obsidian/Books and Movies/Book Log/$year/$month/"

elif [[ $f10 = [0-9]* ]]  && grep -qE '(Medium: Movie|Medium: TV|Medium: Game)' "$1" ; then
        
        mv "$1" "/home/Documents/Obsidian/Books and Movies/Movies and Games/$year/$month/"

fi
```

However, when I try putting that into Shellcommands, e.g. `/bin/bash /home/Documents/Obsidian/Books\ and\ Movies/Scripts/test.sh {{file_name}}` OR `gnome-terminal -- bash -c "bash /home/Documents/Obsidian/Books\ and\ Movies/Scripts/rename-move.sh {{file_name}}; exec bash"`
I keep getting the error that `grep` can't find the file: `grep: test.md: No such file or directory`

I also tried this with `ack` but get the exact same results.

Could you please tell me how to fix this? Thank you!!

**Reply by Taitava** [reply]
Could it be that the working directory is different when you run the script via the _Shell commands_ plugin? The plugin's [default working directory](https://publish.obsidian.md/shellcommands/Basic+usage/Working+directory) is the Obsidian vault's root directory.

Could you try these:
```bash
cd {{folder_path:absolute}}
/bin/bash /home/Documents/Obsidian/Books\ and\ Movies/Scripts/test.sh {{file_name}}
```

Or:
```bash
cd {{folder_path:absolute}}
gnome-terminal -- bash -c "bash /home/Documents/Obsidian/Books\ and\ Movies/Scripts/rename-move.sh {{file_name}}; exec bash"
```

Edit: Oh, you actually solved it already [here](https://github.com/Taitava/obsidian-shellcommands/discussions/369#discussioncomment-6949414).

**Comment by ReaderGuy42** [comment, marked as answer]
I ported the above script back into ShellCommands, and now it works, I think:
```
gnome-terminal -- bash -c "cd {{folder_path:absolute}};  if [ {{file_name}} != 2* ] && grep -qE '(Medium: Book|Medium: Audiobook|Medium: Podcast|Medium: eBook|Medium: GraphicNovel)' {{file_name}}; then mv {{file_name}} /home/Documents/Obsidian/Books\ and\ Movies/Book\ Log/{{date:YYYY}}/{{date:MM}}/{{date:YYYY-MM-DD}}\ {{file_name}}; elif [ {{file_name}} != 2* ] && grep -qE '(Medium: Movie|Medium: TV|Medium: Game)' {{file_name}}; then mv {{file_name}} /home/Documents/Obsidian/Books\ and\ Movies/Movies\ and\ Games/{{date:YYYY}}/{{date:MM}}/{{date:YYYY-MM-DD}}\ {{file_name}}; fi; if [ {{file_name}} = 2* ] && grep -qE '(Medium: Book|Medium: Audiobook|Medium: Podcast|Medium: eBook|Medium: GraphicNovel)' {{file_name}}; then mv {{file_name}} /home/Documents/Obsidian/Books\ and\ Movies/Book\ Log/{{date:YYYY}}/{{date:MM}}/{{file_name}}; elif [ {{file_name}} = 2* ] && grep -qE '(Medium: Movie|Medium: TV|Medium: Game)' {{file_name}}; then mv {{file_name}} /home/Documents/Obsidian/Books\ and\ Movies/Movies\ and\ Games/{{date:YYYY}}/{{date:MM}}/{{file_name}}; fi; exit; exec bash"
```

The `grep` problem I was having seems to be solved by `cd`ing into the `{{folder_path:absolute}}`. Till I tried that I thought that the variable `{{file_name}}` included that, but it seems that was wrong.

Or because it's a lot easier on the eyes, when editing the code, I figured that I can also `cd` into the folder and then launch the external bash script à la `cd {{folder_path:absolute}}; bash "/Documents/Obsidian/Scripts/rename-move.sh" {{file_name}}`

**Reply by Taitava** [reply]
Correct, [`{{file_name}}`](https://publish.obsidian.md/shellcommands/Variables/%7B%7Bfile_name%7D%7D) doesn't contain the folder path part, only the file's name and extension. 👍

**Comment by Taitava** [comment]
> Just aesthetically/visually, it would be easier if your plugin could be written with line breaks in them. Those seem to make the scripts unusable so it seems for now I'm stuck with one-liners. But, as I have no idea how difficult this would be to change, that's fine.

Hmm, I think there a couple of things that could be considered to make your script multiline:
1) As you have your shell commands wrapped in a `gnome-terminal -- bash -c "..."` call, I guess you could try to put a backslash `\` to each line-end when making the script multiline:
    ```bash
    gnome-terminal -- bash -c "case {{yaml_value:Medium}} in Book | Audiobook | Podcast | eBook | GraphicNovel) \
    mv {{file_path:absolute}}  {{vault_path}}/Book\ Log/{{date:YYYY}}/{{date:MM}}/{{date:YYYY-MM-DD}}\ {{file_name}} \
    ;; Movie | TV | Game) \
    mv {{file_path:absolute}}  {{vault_path}}/Movies\ and\ Games/{{date:YYYY}}/{{date:MM}}/{{date:YYYY-MM-DD}}\ {{file_name}};; esac; exit; exec bash"
    ```
I didn't test the script, but it should show the general idea. The backslashes `\` tell Bash to consider the next line belonging to the same statement.

2) I'm not certain why you need `gnome-terminal -- bash -c "..."` in the first place? I'm not on a Linux machine right now, so I can't test, but does `gnome-terminal` open a visual terminal window so you can see the executed shell commands?

**Reply by Taitava** [reply]
> What's more important: how do I tell the shell script to stay with the current note? When I have the above script move my file, Obsidian then opens another note (I can't tell why it chooses that other note). So X is active, I trigger the shell command to move the file, but X should remain active. How can I do this?

Obsidian notices that the file you had open does not exist anymore. The file was renamed or moved, but Obsidian doesn't know the file's new location, so it switches to some other file. There are two ways to overcome this:
1) Use Obsidian URI to open the new file path in Obsidian:
```bash
xdg-open "obsidian://open?path=%2Fhome%2Fuser%2Fmy%20vault%2Fpath%2Fto%2Fmy%20note"
```
A downside of this is that you need to convert all slashes `/` to `%2F` (URL encoding).

2) The _Shell commands_ plugin can open the file for you, if you set your shell command's `stdout` output channel to [Open files](https://publish.obsidian.md/shellcommands/Output+handling/Output+channel+-+Open+files). Then you just need to `echo` the new file path in your script. The downside of this is that you cannot echo anything else, as all output to `stdout` then needs to contain file path(s), nothing else.
Example:
```bash
echo {{vault_path}}/Movies\ and\ Games/{{date:YYYY}}/{{date:MM}}/{{date:YYYY-MM-DD}}\ {{file_name}}:{{caret_position}}
```
An extra feature with this approach is that you can have the caret positioned in the same place where it was before moving/renaming the file. This is accomplished by adding a colon `:` and [`{{caret_position}}`](https://publish.obsidian.md/shellcommands/Variables/%7B%7Bcaret_position%7D%7D) to the end of the file path, so the SC plugin knows to place the caret to the desired place.

**Reply by ReaderGuy42** [reply]
Sorry for the delay. I've set stdout to `Open files` and added the `echo...` thing to my shellcommand. But I have two problems. A) this isn't working:
![image](https://github.com/Taitava/obsidian-shellcommands/assets/71895400/a337bacd-6b6d-488d-8a76-9c87c42be4a9)

Or rather it doesn't _always_ work. It sometimes does move the file but still gives me an error like that.

The second problem is that not all my notes would go into the folder `Movies and Games`. Some would go into `Book Log`, which depends on a frontmatter key. So I would somehow need to add that to the if-statements in my script, but I'm not sure how.

Any advice?
Thanks so much :)

**Reply by Taitava** [reply]
Sorry it took so long for me to reply! Maybe you have already found a solution to this, but in case you have not, here's one attempt:

It seems that in [my previous reply](https://github.com/Taitava/obsidian-shellcommands/discussions/369#discussioncomment-7036730), I have forgotten to take into account that you have two folders (_Movies and Games_ and _Book Log_). I think it's best to create a variable in Bash:

```bash
case {{yaml_value:Medium}}  in

  Book | Audiobook | Podcast | eBook | GraphicNovel)
    TargetFolder="Book Log"
    ;;
  *)
    TargetFolder="Movies and Games"
    ;;
esac
mv {{file_path:absolute}}  {{vault_path}}/"$TargetFolder"/{{date:YYYY}}/{{date:MM}}/{{date:YYYY-MM-DD}}\ {{file_name}} \
echo {{vault_path}}/"$TargetFolder"/{{date:YYYY}}/{{date:MM}}/{{date:YYYY-MM-DD}}\ {{file_name}}:{{caret_position}}
```

I've removed the `gnome-terminal -- bash -c ` part to simplify the script. if you need it, we can add it back (just let me know).
I didn't test the last two lines (`mv` and `echo`), but I tested that the `case` block works and assigns a folder name to the Bash variable depending on the YAML value.

Please let me know, if this works for you, or if there's still some problems. 👍
